# Extra Credit 
# 3. Create a tutorial on an big data algorthim_Random Forest in Python

Random Forest Algorithm: 

Random forest is a type of supervised machine learning algorithm based on ensemble learning. Ensemble learning is a type of learning where you join different types of algorithms or same algorithm multiple times to form a more powerful prediction model. The random forest algorithm combines multiple algorithm of the same type i.e. multiple decision trees, resulting in a forest of trees, hence the name "Random Forest". 

Problem statement:

The problem here is to predict the gas consumption (in millions of gallons) in 48 of the US states based on petrol tax (in cents), per capita income (dollars), paved highways (in miles) and the proportion of population with the driving license.

In [2]:
import pandas as pd  
import numpy as np  

In [4]:
# Importing Dataset
dataset = pd.read_csv('petrol_consumption.csv') 
dataset.head() 

Petrol_tax  Average_income  Paved_Highways  Population_Driver_licence(%)  \
0         9.0            3571            1976                         0.525   
1         9.0            4092            1250                         0.572   
2         9.0            3865            1586                         0.580   
3         7.5            4870            2351                         0.529   
4         8.0            4399             431                         0.544   

   Petrol_Consumption  
0                 541  
1                 524  
2                 561  
3                 414  
4                 410

In [5]:
# Preparing Data For Training
# divide data into 'attributes' and 'label' sets
X = dataset.iloc[:, 0:4].values  
y = dataset.iloc[:, 4].values  

In [6]:
# divide the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)  

In [7]:
# Feature Scaling
# for instance the Average_Income field has values in the range of thousands while Petrol_tax has values in range of tens. 
# it would be beneficial to scale our data 
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()  
X_train = sc.fit_transform(X_train)  
X_test = sc.transform(X_test)  

In [8]:
# Training the Algorithm
 
from sklearn.ensemble import RandomForestRegressor

regressor = RandomForestRegressor(n_estimators=20, random_state=0)  
regressor.fit(X_train, y_train)  
y_pred = regressor.predict(X_test)  

In [9]:
# Evaluating the Algorithm
from sklearn import metrics

print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))  # compute mean absolute error
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))  # compute mean squared error
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))  # compute root mean squared error


('Mean Absolute Error:', 51.76500000000001)
('Mean Squared Error:', 4216.166749999999)
('Root Mean Squared Error:', 64.93201637097064)


Discussion for the result:

With 20 trees, the root mean squared error is 64.93 which is greater than 10 percent of the average petrol consumption i.e. 576.77. This may indicate, among other things, that we have not used enough estimators.

Reference:
https://stackabuse.com/random-forest-algorithm-with-python-and-scikit-learn/